62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+            assert((candidatesSize % 2u)==0u);
+            for (auto i=0u; i< candidatesSize;)
+            {
+                auto left = activeCandidates[i++];
+                auto right = activeCandidates[i++];
+
+                CurveHatchBox curveBox;
+
+				left.t_end = left.originalBezier->intersectOrtho(newMajor, major);
+				right.t_end = right.originalBezier->intersectOrtho(newMajor, major);
+                auto splitCurveMin = left.getSplitCurve();
+                auto splitCurveMax = right.getSplitCurve();
+				assert(splitCurveMin.evaluateBezier(0.0)[major] <= splitCurveMin.evaluateBezier(1.0)[major]);
+				assert(splitCurveMax.evaluateBezier(0.0)[major] <= splitCurveMax.evaluateBezier(1.0)[major]);
+
+                auto curveMinAabb = splitCurveMin.getBezierBoundingBoxMinor();
+                auto curveMaxAabb = splitCurveMax.getBezierBoundingBoxMinor();
+                curveBox.aabbMin = float64_t2(std::min(curveMinAabb.first.x, curveMaxAabb.first.x), lastMajor);
+                curveBox.aabbMax = float64_t2(std::max(curveMinAabb.second.x, curveMaxAabb.second.x), newMajor);
+
+				if (isCurrentDebugStep)
+				{
+					drawDebugBezier(splitCurveMin, float64_t4(1.0, 0.0, 0.0, 1.0));
+					drawDebugBezier(splitCurveMax, float64_t4(0.0, 1.0, 0.0, 1.0));
+
+					//printf(std::format("Split curve min ({}, {}), ({}, {}), ({}, {}) max ({}, {}), ({}, {}), ({}, {})\n",
+					//	splitCurveMin.p[0].x, splitCurveMin.p[0].y,
+					//	splitCurveMin.p[1].x, splitCurveMin.p[1].y,
+					//	splitCurveMin.p[2].x, splitCurveMin.p[2].y,
+					//	splitCurveMax.p[0].x, splitCurveMax.p[0].y,
+					//	splitCurveMax.p[1].x, splitCurveMax.p[1].y,
+					//	splitCurveMax.p[2].x, splitCurveMax.p[2].y
+					//).c_str());
+
+				}
+
+                //if (debugOutput)
+                //{
+                //    drawDebugLine(float64_t2(curveBox.aabbMin.x, curveBox.aabbMin.y), float64_t2(curveBox.aabbMax.x, curveBox.aabbMin.y), float32_t4(0.0, 0.3, 0.0, 0.1));
+                //    drawDebugLine(float64_t2(curveBox.aabbMax.x, curveBox.aabbMin.y), float64_t2(curveBox.aabbMax.x, curveBox.aabbMax.y), float32_t4(0.0, 0.3, 0.0, 0.1));
+                //    drawDebugLine(float64_t2(curveBox.aabbMin.x, curveBox.aabbMax.y), float64_t2(curveBox.aabbMax.x, curveBox.aabbMax.y), float32_t4(0.0, 0.3, 0.0, 0.1));
+                //    drawDebugLine(float64_t2(curveBox.aabbMin.x, curveBox.aabbMin.y), float64_t2(curveBox.aabbMin.x, curveBox.aabbMax.y), float32_t4(0.0, 0.3, 0.0, 0.1));
+                //}
+                // Transform curves into AABB UV space and turn them into quadratic coefficients
+                // TODO: the split curve should already have the quadratic bezier as
+                // quadratic coefficients
+                // so we wont need to convert here
+                auto transformCurves = [](Hatch::QuadraticBezier bezier, float64_t2 aabbMin, float64_t2 aabbMax, float64_t2* output) {
```

Comment:
1. you'll be changing `Hatch::QuadraticBezier` to `hlsl::shapes::QuadraticBezier` as per previous comments
2. instead of manually computing a,b,c do ` Quadratic::constructFromBezier(QuadraticBezier<float_t> curve)`

Reply:
look in discord, importing this hsa compilation issues